File: buildSrc/src/main/groovy/com/engineer/plugin/transforms/times/CatClassVisitor.java
package com.engineer.plugin.transforms.times;

import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;


public class CatClassVisitor extends ClassVisitor {

    private String mClassName;

    public CatClassVisitor(ClassVisitor classVisitor) {
        super(Opcodes.ASM6, classVisitor);
    }

    @Override
    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
        super.visit(version, access, name, signature, superName, interfaces);
        this.mClassName = name;
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
        MethodVisitor methodVisitor = super.visitMethod(access, name, desc, signature, exceptions);
        methodVisitor = new CatMethodVisitor(Opcodes.ASM6, methodVisitor, access, mClassName, name, desc);
        return methodVisitor;
    }
}

File: buildSrc/src/main/kotlin/com/engineer/plugin/transforms/cost/CatClassVisitor.kt
package com.engineer.plugin.transforms.cost

import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes

class CatClassVisitor(classVisitor: ClassVisitor) : ClassVisitor(Opcodes.ASM6, classVisitor) {

    private lateinit var mClassName: String

    override fun visit(
        version: Int, access: Int, name: String,
        signature: String?, superName: String?, interfaces: Array<String>?
    ) {
        super.visit(version, access, name, signature, superName, interfaces)
        mClassName = name
    }

    override fun visitMethod(
        access: Int, name: String, desc: String?,
        signature: String?, exceptions: Array<String>?
    ): MethodVisitor {
        var methodVisitor = super.visitMethod(access, name, desc, signature, exceptions)

        methodVisitor =
            CatMethodVisitor(
                Opcodes.ASM6,
                methodVisitor,
                access,
                mClassName,
                name,
                desc
            )
        return methodVisitor
    }
}